lib/aws.js

```diff
@@ -133,6 +133,54 @@ module.exports.uploadDirectoryToS3Async = async function(s3Bucket, s3Path, local
 };
 module.exports.uploadDirectoryToS3 = util.callbackify(this.uploadDirectoryToS3Async);
 
+/**
+ * Download a file or directory from S3.
+ * 
+ * @param {string} s3Bucket - The S3 bucket name.
+ * @param {string} s3Path - The S3 source path.
+ * @param {string} localPath - The local target path.
+ * @param {object?} options - Optional parameters, including owner and group (optional, defaults to {}).
+ */
+module.exports.downloadFromS3Async = async function (s3Bucket, s3Path, localPath, options={}) {
+    if (localPath.endsWith('/')) {
+        debug(`downloadFromS3: bypassing S3 and creating directory localPath=${localPath}`);
+        await fs.promises.mkdir(localPath, { recursive: true });
+        if (!isNaN(options.owner) && !isNaN(options.group)) await fs.promises.chown(localPath, options.owner, options.group);
+        return;
+    }
+
+    debug(`downloadFromS3: creating containing directory for file localPath=${localPath}`);
+    await fs.promises.mkdir(path.dirname(localPath), { recursive: true });
+    if (!isNaN(options.owner) && !isNaN(options.group)) await fs.promises.chown(path.dirname(localPath), options.owner, options.group);
```

Comment:
Same `isNaN()` change. Do we need to worry about chowning intermediate paths? For example, if we make `a/b/c/d`, do we just need to chown `d` or should we also chown `c`, `b`, etc?

Reply:
It turns out nested dirs will currently break the workspace launch (usually). Nested dirs used to work at some point, so I'm not sure when exactly this regressed.

```
info: Launching workspace-1-1 (useInitialZip=true)
  prairielearn:interface Bootstrapping workspace with initial.zip +0ms
  prairielearn:workspace Setting workspaces.message to 'Loading initial files' +0ms
  prairielearn:interface Downloading s3Path=workspace-1-1/initial.zip to zipPath=/workspace_host_zips/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba-initial.zip +2ms
  prairielearn:aws downloadFromS3: creating containing directory for file localPath=/workspace_host_zips/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba-initial.zip +0ms
  prairielearn:interface Making directory /jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba +50ms
  prairielearn:interface Preparing unzip jobs +13ms
info: Watching directory add /jobs/workspaces
  prairielearn:interface Unzipping /workspace_host_zips/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba-initial.zip to /jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba +20ms
  prairielearn:interface Making directory /jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a/ +1ms
  prairielearn:interface Making directory /jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a/b/ +1ms
  prairielearn:interface Making directory /jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a/b/c/ +0ms
  prairielearn:interface Extracting file /jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a/b/c/data.csv +1ms
  prairielearn:interface Chowning entryPath=/jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a/b/c/data.csv +1ms
  prairielearn:interface Extracting file /jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/fibonacci.py +1ms
  prairielearn:interface Chowning entryPath=/jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/fibonacci.py +1ms
  prairielearn:interface Chowning entryPath=/jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a/ +2ms
  prairielearn:interface Chowning entryPath=/jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a/b/ +3ms
info: Watching directory add /jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba
  prairielearn:workspace Setting workspaces.state='stopped', workspaces.message='Error! Click "Reboot" to try again. Detail: Error: ENOENT: no such file or directory, chown '/jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a/b/c/data.csv'' +101ms
events.js:292
      throw er; // Unhandled 'error' event
      ^

Error: ENOENT: no such file or directory, open '/jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a/b/c/data.csv'
Emitted 'error' event on WriteStream instance at:
    at errorOrDestroy (internal/streams/destroy.js:108:12)
    at WriteStream.onerror (_stream_readable.js:752:7)
    at WriteStream.emit (events.js:315:20)
    at WriteStream.EventEmitter.emit (domain.js:483:12)
    at /PrairieLearn/node_modules/graceful-fs/graceful-fs.js:211:14
    at /PrairieLearn/node_modules/graceful-fs/graceful-fs.js:241:16
    at FSReqCallback.oncomplete (fs.js:156:23) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a/b/c/data.csv'
}
```
```
[root@81dbee9b9013 PrairieLearn]# find /jobs/workspaces/ -print
/jobs/workspaces/
/jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba
/jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a
/jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a/b
/jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/a/b/c
/jobs/workspaces/workspace-9220926b-6e02-45c7-9d40-bfbc68f451ba/fibonacci.py
```

Reply:
Actually I'm not sure it's related to nested dirs. It seems to be an unzipping issue overall.

Reply:
It seems to have something to do with the parallelized unzip jobs (for `initial.zip`), so I reverted that change. But with the original unzip code, I'm still getting inconsistent unzipping of `initial.zip`.

Reply:
I'm happy to go back to serial zip/unzip. It's only a small speedup anyway and it's possible that the `unzipper` library doesn't play well with parallel unzip.

Reply:
I haven't investigated in detail, but in the [Quick Examples](https://www.npmjs.com/package/unzipper#quick-examples) section of the `unzipper` docs they say "Extract emits the 'close' event once the zip's contents have been fully extracted to disk." We don't currently wait for this, so I suspect we are returning before the individual file unzips have actually occurred.

Reply:
I think it might be as simple as doing:
```
await entry.pipe(fs.createWriteStream(entryPath)).promise();
```

Reply:
```diff
-             entry.pipe(fs.createWriteStream(entryPath));
+             await entry.pipe(fs.createWriteStream(entryPath)).promise();
```

^ that still didn't work on my end, but this `stream.Transform` pipe is working consistently so far on my end:

https://www.npmjs.com/package/unzipper#parse-zip-by-piping-entries-downstream

Not sure if there's a cleaner way, but I pushed this piping/transform version for now.

Reply:
I believe this comment from the `unzipper` author confirms that the basic async iterator won't work for our use case (parallel or not). From https://github.com/ZJONSSON/node-unzipper/issues/138#issuecomment-506376795:

> The problem is that the pipe chain finishes when the zip file has been read, not when the output files have been written. For small files this is roughly the same thing, but for large files there can be a some lag.

He suggests:

> If you are using unzipper.Parse() one way to tackle this is to use .pipe into a transform stream that only performs callback when each file has been written. You can see some examples in the readme here: https://www.npmjs.com/package/unzipper#parse-zip-by-piping-entries-downstream

So this code is working on my end and pushed upstream. I don't know how to get rid of the ugly callback+async mixing here.

https://github.com/PrairieLearn/PrairieLearn/blob/2a700d7c4a61ff51c9367077f79c76eada98b625/workspace_host/interface.js#L795-L817

Reply:
Reverted and moved to #3146